### PR TITLE
Try to detect unicode normalization issues in column names

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2901

### CODE REVIEW

**Comment by bkamins** [comment]
The only problem is that this fix does not resolve the OP issue as still:
```
julia> d = DataFrame("power_µW" => [1,2,3])
3×1 DataFrame   
 Row │ power_µW 
     │ Int64    
─────┼──────────
   1 │        1 
   2 │        2 
   3 │        3 

julia> d.power_µW
ERROR: ArgumentError: column name :power_μW not found in the data frame; existing most similar names are: :power_µW
```
as this kind of change is not covered by `Unicode.normalize`.

**Comment by nalimilan** [comment]
What's annoying is that we don't give any concrete solution when the parser replaced some chars with others during normalization. We should probably wait until @stevengj's PR is merged.

**Comment by stevengj** [comment]
If there is no ambiguity, I'm not quite clear on why you are throwing an error rather than giving the equivalent (modulo normalization) column?

**Comment by nalimilan** [comment]
Also performance will be worse when we need to check all column names, so better signal the problem to the user than silently being slower.

**Comment by stevengj** [comment]
> So that would mean doing first a lookup of raw names, and if that fails a lookup on normalized names?

Yes, but furthermore in `getproperty` I think you should only look up normalized names, and should assume that the Symbol argument has already been normalized by Julia.

**Comment by stevengj** [comment]
@bkamins, it's already been decided that `Symbol("...")` constructs a programmatic symbol that may not be type-able by the user, and is not normalized.

**Comment by bkamins** [comment]
> it's already been decided that `Symbol("...")` constructs a programmatic symbol that may not be type-able by the user, and is not normalized.

Yes. I understand this. What I do not understand is why you think that:

```
julia> nt = (; Symbol("noe\u0308l") => ["noel"])
(noël = ["noel"],)

julia> julia> nt.noël # autocompleted
ERROR: type NamedTuple has no field noël
```

is OK (as this is what I understand your comment implies).

While you think that this behavior should be changed:

```
julia> df = DataFrame(nt)
1×1 DataFrame  
 Row │ noël   
     │ String  
─────┼──────── 
   1 │ noel    

julia> df.noël
ERROR: ArgumentError: column name :noël not found in the data frame; existing most similar names are: :noël
```

Or putting it in different words why:

> you could still allow such names in the non-normalized lookup, but you would omit names with normalization conflicts from the normalized lookup table—hence only a normalized lookup of an ambiguous normalized name (e.g. via `getproperty`) would be an error.

should apply to `DataFrame`, but should not apply to `NamedTuple`?

The reason I am pressing for clarifying this (and avoid just implementing it) is that there is an intended duality between these two types. `NamedTuple` of vectors of equal length and `DataFrame` should be useable in exactly the same way with the exception that `NamedTuple` is type stable and `DataFrame` is not type stable. This e.g. has a consequence that internally in DataFrames.jl we construct such `NamedTuples` to improve performance of some operations via a function barrier and having such discrepancy would be problematic.

File: src/other/index.jl
Comment (bkamins) [inline src/other/index.jl:293]:
```suggestion
# see JuliaLang/julia#42561; we keep our own list to ensure things work consistently on earlier versions of Julia
```

File: src/other/index.jl
Comment (bkamins) [reply src/other/index.jl]: I have re-implemented it to provide a simpler logic.

File: src/other/index.jl
Comment (bkamins) [reply src/other/index.jl]: I have made three separate error messages (tests cover all options).

**Comment by bkamins** [comment]
@nalimilan - I have updated the PR. The current approach is not to make any breaking changes, but just give more informative error messages.

In the future if we get reports from users that what we do now is problematic we could go back to discuss changes that @stevengj proposed (however, maybe adding special errors will be enough and this way we do not make breaking changes).

**Comment by stevengj** [comment]
> Note that it is even super hard to type in "power_µW" in Julia 

Not on a Mac, where option-m is `µ` (micro).

**Comment by bkamins** [comment]
> people kept confusing `µ` (micro) and `μ` (mu) in real code.

This is indeed a valid issue (it is interesting how such things influence experience; in my region Mac is quite rare).

However, I would still recommend that we take the path:
* first improve the error message in 1.3 release and wait for the feedback;
* based on the feedback decide if we want to make a breaking change.

File: src/other/index.jl
Comment (nalimilan) [inline src/other/index.jl:347]: This would be more explicit IMO, as otherwise "normalized passed column name" could be understood as "the passed name is already normalized".
```suggestion
                                        "data frame. However, after applying Unicode " *
                                        "normalization to it, the passed column name " *
                                        "matches a column name found in the data frame." *
                                        "You can use the `Unicode.normalize` " *
                                        "function to normalize the passed column name."))
```

File: src/other/index.jl
Comment (nalimilan) [inline src/other/index.jl:353]:
```suggestion
                                        "data frame. However, there is a match " *
                                        "after applying Unicode normalization both " *
                                        "to the passed column names " *
                                        "and to the column names of the data frame. " *
                                        "In the call both passed column and " *
```

File: src/other/index.jl
Comment (nalimilan) [inline src/other/index.jl:362]:
```suggestion
                                        "function to normalize the passed column name."))
```

File: src/other/index.jl
Comment (nalimilan) [inline src/other/index.jl:369]:
```suggestion
                                    "data frame. However, the passed column name " *
                                    "matches a column name found in the data frame " *
                                    "after applying Unicode normalization to the latter." *
                                    "It is recommended to use " *
```

File: src/other/index.jl
Comment (nalimilan) [inline src/other/index.jl:319]:
```suggestion
                                "(codepoint: $(UInt32(c2))) is used instead of $c1 " *
```

File: src/other/index.jl
Comment (nalimilan) [inline src/other/index.jl:324]: Now that https://github.com/JuliaLang/julia/pull/42561has been merged, could we print the command that fixes column names on Julia ≥ 1.8?

**Comment by nalimilan** [comment]
> > people kept confusing `µ` (micro) and `μ` (mu) in real code.
> 
> This is indeed a valid issue (it is interesting how such things influence experience; in my region Mac is quite rare).

FWIW, on French PC keyboards, we also have micro available (with Shift+a key close to Return). Apparently some variants of US and Polish keyboards also support it via AltGr+M. That's indeed probably the main reason why people get confused.

File: src/other/index.jl
Comment (bkamins) [reply src/other/index.jl]: OK - I have changed the error message under `v"1.8-"`.

File: src/other/index.jl
Comment (nalimilan) [inline src/other/index.jl:318]: How about a more direct suggestion like this?
```suggestion
                suffix = "normalize column names of the data frame by calling " *
                         "`using Unicode; rename!(n -> Unicode.normalize(n, chartransform=Unicode.julia_chartransform), df)`."
```

Comment (bkamins) [reply src/other/index.jl:318]: This is a half-measure solution (just like the one below earlier). I will propose a better error (as we need to signal the user if the passed name or the name in data frame was incorrect)

Comment (bkamins) [reply src/other/index.jl]: @nalimilan In the end I have reverted the simpler error message. The issue is that the number of potential cases we would have to handle is too large to warrant covering all of them given how rare the case is. I now just write which character should be replaced by which.

Just to give you a flavor of what happens:

`"\u0387"` after standard normalization via `Unicode.normalize` becomes `"\u00B7"` which after Julia-specific normalization (`chartransform=Unicode.julia_chartransform`) becomes "\u22C5".

I think that the problematic case is rare enough that we can just write what characters do not match and let the user decide how to handle this.

Comment (nalimilan) [reply src/other/index.jl]: OK, I've lost track of all the possible cases to handle...

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**
